File: EvaluateDivision2.py
class Solution:
    def calcEquation(self, equations, values, queries):
        """
        :type equations: List[List[str]]
        :type values: List[float]
        :type queries: List[List[str]]
        :rtype: List[float]
        """
        class Node:
            def __init__(self):
                self.parent = self
                self.value = 0

        def findParent(node):
            if node.parent == node:
                return node
            node.parent = findParent(node.parent)
            return node.parent

        def unionNodes(node1, node2, value):
            parent1, parent2 = findParent(node1), findParent(node2)
            ratio = node2.value * value / node1.value
            for k, n in m.items():
                if findParent(n) == parent1:
                    n.value *= ratio
            parent1.parent = parent2

        m = {}
        for i, equation in enumerate(equations):
            if equation[0] not in m and equation[1] not in m:
                m[equation[0]] = Node()
                m[equation[1]] = Node()
                m[equation[1]].value = 1
                m[equation[0]].value = values[i]
                m[equation[0]].parent = m[equation[1]]
            elif equation[0] not in m:
                m[equation[0]] = Node()
                m[equation[0]].value = m[equation[1]].value * values[i]
                m[equation[0]].parent = m[equation[1]]
            elif equation[1] not in m:
                m[equation[1]] = Node()
                m[equation[1]].value = m[equation[0]].value / values[i]
                m[equation[1]].parent = m[equation[0]]
            else:
                unionNodes(m[equation[0]], m[equation[1]], values[i])

        result = [-1] * len(queries)
        for i, query in enumerate(queries):
            if query[0] not in m or query[1] not in m or findParent(m[query[0]]) != findParent(m[query[1]]):
                continue
            result[i] = m[query[0]].value / m[query[1]].value

        return result


if __name__ == '__main__':
    n = int(input())
    equations = []
    for _ in range(n):
        equations.append([x for x in input().split()])
    values = [float(v) for v in input().split()]
    m = int(input())
    queries = []
    for _ in range(m):
        queries.append([x for x in input().split()])

    solution = Solution()
    print(solution.calcEquation(equations, values, queries))

File: EvaluateDivision2.cpp
#include<iostream>
#include<vector>
#include<unordered_map>
#include<unordered_set>

using namespace std;

class Solution {
public:
    vector<double> calcEquation(vector<pair<string, string> > equations, vector<double>& values, vector<pair<string, string> > queries) {
    	
	    unordered_set<string> s;
	    for(auto equation : equations) {
	    	s.insert(equation.first);
	    	s.insert(equation.second);
		}
		
		vector<double> result(queries.size(), -1);
		for(int i = 0; i < queries.size(); i++) {
			
			if(s.find(queries[i].first) != s.end() && s.find(queries[i].second) != s.end()) {
				vector<string> visited;
				result[i] = calcEquation(equations, values, queries[i].first, queries[i].second, visited);
			}
			
		}
		
		return result;
	    
    }
    
    double calcEquation(vector<pair<string, string> > equations, vector<double>& values, string x, string y, vector<string>& visited) {
    	if(x == y) return 1;
    	for(int i = 0; i < equations.size(); i++) {
    		if(x == equations[i].first && y == equations[i].second) return values[i];
    		if(x == equations[i].second && y == equations[i].first) return 1 / values[i];
		}
		
		for(int i = 0; i < equations.size(); i++) {
			if(x == equations[i].first) {
				bool is_ok = true;
				for(int j = 0; j < visited.size(); j++) {
					if(visited[i] == equations[i].second) {
						is_ok = false;
						break;
					}
				}
				if(is_ok) {
					visited.push_back(x);
					double t = calcEquation(equations, values, equations[i].second, y, visited);
					if(t != -1) return t * values[i];
				}
			}else if(x == equations[i].second) {
				bool is_ok = true;
				for(int j = 0; j < visited.size(); j++) {
					if(visited[i] == equations[i].first) {
						is_ok = false;
						break;
					}
				}
				if(is_ok) {
					visited.push_back(x);
					double t = calcEquation(equations, values, equations[i].first, y, visited);
					if(t != -1) return t / values[i];
				}
			}
		}
		
		return -1;
	}
};

int main() {
	
	int n;
	cin>>n;
	
	vector<pair<string, string> > equations;
	for(int i = 0; i < n; i++) {
		string s1, s2;
		cin>>s1>>s2;
		equations.push_back(make_pair(s1, s2));
	}
	
	vector<double> values;
	for(int i = 0; i < n; i++) {
		double v;
		cin>>v;
		values.push_back(v);
	}
	
	cin>>n;
	vector<pair<string, string> > queries;
	for(int i = 0; i < n; i++) {
		string s1, s2;
		cin>>s1>>s2;
		queries.push_back(make_pair(s1, s2));
	}
	
	Solution *solution = new Solution();
	vector<double> result = solution->calcEquation(equations, values, queries);
	
	for(int i = 0; i < result.size(); i++) cout<<result[i]<<"  ";
	
	return 0;
	
}
